File: app/src/main/java/com/example/chap16recyclerviewexample/MainActivity.java
package com.example.chap16recyclerviewexample;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.os.Bundle;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private RecyclerAdapter adapter;
    private ArrayList<String> difficultyList = new ArrayList<>();
    private ArrayList<String> descriptionList = new ArrayList<>();
    private ArrayList<Integer> imageList = new ArrayList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        recyclerView = findViewById(R.id.recyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(MainActivity.this));

        addArray();

        adapter = new RecyclerAdapter(difficultyList, descriptionList, imageList, MainActivity.this);
        recyclerView.setAdapter(adapter);
    }

    private void addArray() {
        difficultyList.add("Easy");
        difficultyList.add("Asian");
        difficultyList.add("Chinese");
        difficultyList.add("Emotional Damage");
        descriptionList.add("You select easy mode, too easy lah.");
        descriptionList.add("You are a FAILURE!");
        descriptionList.add("Neighbor's kid do better.");
        descriptionList.add("I will send you to Jesus.");
        imageList.add(R.drawable.easy);
        imageList.add(R.drawable.asian);
        imageList.add(R.drawable.chinese);
        imageList.add(R.drawable.emotional_damage);
    }
}

File: app/src/main/java/com/example/chap6/MainActivity.kt
package com.example.chap6

import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.ArrayAdapter
import android.widget.GridView
import android.widget.ListView
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView

class MainActivity : AppCompatActivity() {

    private lateinit var myRecyclerView: RecyclerView
    var difficulty = ArrayList<String>()
    var description = ArrayList<String>()
    var image = ArrayList<Int>()
    private lateinit var adapter: DifficultiesAdapter

    private fun fillArray() {
        difficulty.add("Easy")
        difficulty.add("Asian")
        difficulty.add("Chinese")
        difficulty.add("Emotional Damage")
        description.add("You select easy mode, too easy lah.")
        description.add("You are a FAILURE!")
        description.add("Neighbor's kid do better.")
        description.add("I will send you to Jesus.")
        image.add(R.drawable.easy)
        image.add(R.drawable.asian)
        image.add(R.drawable.chinese)
        image.add(R.drawable.emotional_damage)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        fillArray()

        myRecyclerView = findViewById(R.id.recyclerView1)
        myRecyclerView.layoutManager = LinearLayoutManager(this) // This makes your layout become LinearLayout.

        adapter = DifficultiesAdapter(difficulty, description, image, this)
        myRecyclerView.adapter = adapter

    }
}